### PR TITLE
Volatile DB: use modifyWithTempRegistry

### PR DESCRIPTION
**Pull request by mrBliss**
* Rewrite `tryVolDB` and `wrapFsError`. Follow the approach taken in the ImmutableDB.
* Split off state-related functions in `Impl.State`.
* Move modules under `Impl`.
* Introduce `withState` and use it.
* Rename things to be consistent with the ImmutableDB.
* Use `modifyWithTempRegistry` to update the state.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+    -- close doesn't take the state that @open@ returned, because the state
+    -- may have been updated by someone else since we got it (remember we're
+    -- using 'readMVar' here, 'takeMVar'). So we need to get the most recent
+    -- state anyway.
+    close :: ExitCase (Either VolatileDBError r)
+          -> m ()
+    close ec = case ec of
+      ExitCaseAbort                               -> return ()
+      ExitCaseException _ex                       -> return ()
```
Comment (kderme) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs]: At `modifyOpenState` we close the db in this case. Why is it different here?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+-- In case an 'UnexpectedError' is thrown while the action is being run, the
+-- database is closed to prevent further appending to a database in a
+-- potentially inconsistent state. All other exceptions will leave the
+-- database open.
+withOpenState
+  :: forall blockId m r. (HasCallStack, IOLike m)
+  => VolatileDBEnv m blockId
+  -> (forall h. HasFS m h -> OpenState blockId h -> m r)
+  -> m r
+withOpenState VolatileDBEnv {hasFS = hasFS :: HasFS m h, varInternalState} action = do
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs]: Note `(tryVolDB . access)`, that's where the wrapping is done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs [hunk 3]
```diff
@@ -174,50 +132,52 @@ openDB :: ( HasCallStack
...
             -> m ()
 closeDBImpl VolatileDBEnv { varInternalState, tracer, hasFS = HasFS {..} } = do
-    mbInternalState <- swapMVar varInternalState VolatileDbClosed
+    mbInternalState <- swapMVar varInternalState DbClosed
     case mbInternalState of
-      VolatileDbClosed -> traceWith tracer DBAlreadyClosed
-      VolatileDbOpen InternalState{..} ->
+      DbClosed -> traceWith tracer DBAlreadyClosed
+      DbOpen OpenState{..} ->
         wrapFsError $ hClose currentWriteHandle
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs]: Good idea.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs]: In `modifyOpenState`, we might have written something to disk, so we are no longer sure that our in-memory state matches our on-disk state. We close as a safety measure. Note that we expect `ExitCaseException` mainly to be triggered by async exceptions, in which case we should shut down the database anyway. This is related to #1681.

In `withOpenState`, we are not modifying the state, so an async exception must not and *should* not close the database, as the per-connection threads reading from the VolatileDB *can* be killed.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+        unless (dbIsOpen st') $ closeOpenHandles hasFS ost
+      where
+        st' = case ec of
+          ExitCaseSuccess ost' -> DbOpen ost'
+
+          -- When something goes wrong, close the VolatileDB for safety.
+          -- Except for user errors, because they stem from incorrect use of
+          -- the VolatileDB.
+          --
+          -- NOTE: we only modify the VolatileDB in a background thread of
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs]: You're right, updated.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs [hunk 5]
```diff
@@ -243,10 +207,10 @@ getBlockComponentImpl
   -> blockId
   -> m (Maybe b)
 getBlockComponentImpl env blockComponent blockId =
-    modifyState env $ \hasFS st@InternalState { currentRevMap } ->
+    withOpenState env $ \hasFS OpenState { currentRevMap } ->
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs]: Right, I forgot about that :flushed:. I'm going to keep it simple for now and make it locking (`takeMVar`) again.

We can revisit this in #1816.

**Pull request merged**